README.md

```diff
@@ -161,6 +161,61 @@ auto_ssl:set("allow_domain", function(domain, auto_ssl, ssl_options, renewal)
 end)
 ```
 
+#### `get_failures`
+
+The optional `get_failures` function accepts a domain name argument, and can be used to retrieve statistics about failed certificate requests concerning the domain. The function will return a table with fields `first` (timestamp of first failure encountered), `last` (timestamp of most recent failure encountered), `num` (number of failures). The function will instead return `nil` if no error has been encountered.
+
+Note: the statistics are only kept for as long as the nginx instance is running. There is no sharing across multiple servers (as in a load-balanced environment) implemented.
+
+To make use of the `get_failures` function, add the following to the `http` configuration block:
+
+```nginx
+  lua_shared_dict auto_ssl_failures 1m;
+```
+
+When this shm-based dictionary exists, `lua-resty-auto-ssl` will use it to update a record it keeps for the domain whenever a Let's Encrypt certificate request fails (for both new domains, as well as renewing ones). When a certificate request is successful, `lua-resty-auto-ssl` will delete the record it has for the domain, so that future invocations will return `nil`.
+
+The `get_failures` function can be used inside `allow_domain` to implement per-domain rate-limiting, and similar rule sets.
+
+*Example:*
+
+```lua
+auto_ssl:set("allow_domain", function(domain, auto_ssl, ssl_options, renewal)
+  local failures = auto_ssl:get_failures(domain)
+  -- only attempt one certificate request per hour
+  if not failures or 3600 < ngx.now() - failures["last"] then
+    return true
+  else
+    return false
+  end
+end)
+```
+
+#### `track_failure`
+
+The optional `track_failure` function accepts a domain name argument and records a failure for this domain. This can be used to avoid repeated lookups of a domain in `allow_domain`.
+
+*Example:*
+
+```lua
+auto_ssl:set("allow_domain", function(domain, auto_ssl, ssl_options, renewal)
+  local failures = auto_ssl:get_failures(domain)
+  -- only attempt one lookup or certificate request per hour
+  if failures and ngx.now() - failures["last"] <= 3600 then
+    return false
+  end
+
+  local allow
+  -- (external lookup to check domain, e.g. via http)
+  if not allow then
+    auto_ssl:track_failure(domain)
+    return false
+  else
+    return true
+  end
+end)
+```
+
```

Comment:
This `get_failures` and `track_failure` logic to help throttle initial registrations is definitely a flexible solution that allows for some nifty logic in the `allow_domain` callback. I think I see how all this fits together (nice!), but I guess I'm just mulling over other options to deal with the same issue. A couple things come to mind.

1. You commented:

    > The one slightly ambiguous part IMHO: we're using 6e1ea73 together with b1f9715, a change to check if we already have a cert in permanent storage before calling allow_domain, which was previously rejected during review. Without the latter patch, you might have a situation in a multi-server setup where one server serves a certificate for a domain, while another still returns the fallback certificate because of per-domain throttling. With the latter patch, auto-ssl will always return a certificate any of the servers has been able to get one.

    The ordering issues around this does make me wondering about whether `allow_domain` is the right place to implement this logic. Because in a way, this seems like a slightly different concept than what `allow_domain` as originally intended to do. Perhaps we could introduce a different callback (eg, `throttle_domain`) that would get called after the storage call (eg, slide it into where you had proposed `allow_domain` get moved in https://github.com/auto-ssl/lua-resty-auto-ssl/commit/b1f9715dd24a4ad4c1204d638fac10b2d0d23830)? That would keep `allow_domain` being called before the storage call, but then allow this logic to be handled separately.
2. I'm also wondering if we could integrate this logic in a more built-in fashion so that users don't have to implement custom `allow_domain` (or other callbacks) to handle this. Similar to how `renewals_per_hour` is just a simple global setting, I'm wondering if we could simplify this in some fashion and integrate this logic directly into codebase, rather than relying on the callback. Or do your use-cases need more custom or per-domain logic for this type of thing?

I'd need to spend some more time investigating the feasibility of these ideas (and I realize you might have different use-cases), but these are just some initial thoughts.

Reply:
Thank you for your review and thoughts, @GUI.

To your points:

1. Separate callbacks would definitely work for us. We don't mind the Redis lookup to see if a certificate is in storage or not, but to accommodate users that do, having two makes a lot of sense to me IMHO. (perhaps `_pre` and `_post`?)

Part of my thinking behind adding this to the implementer's "toolbox" in `allow_domain` was the realization that one could use the `track_failure(domain)` mechanism also for caching any "regular" lookup that `allow_domain` would do, if this is desired. (Example: we have to do a HTTP call to an internal web API to decide whether we should allow a new (sub-) domain)

2. I do see beauty of having a simple setting, and have everything be handled behind the scenes by `lua-resty-auto-ssl`.

To be specific, we're currently using this exponential backoff algorithm in our setup. Unsure if something like this could be generalized to work for other users (without a callback?):

      local failures = auto_ssl:get_failures(domain)
      if failures then
        -- 1 failures: next attempt after 3.75 min
        -- 2 failures: next attempt after 7.5 min
        -- 3 failures: next attempt after 15 min
        -- 4 failures: next attempt after 30 min
        -- 5 failures: next attempt after 60 min
        -- ...
        -- 10+ failures: next attempt after 24 hours
        local backoff = math.min(225 * 2 ^ (failures["num"] - 1), 86400)
        local elapsed = ngx.now() - failures["last"]
        if elapsed < backoff then
          ngx.log(ngx.NOTICE, domain .. ": " .. failures["num"] .. " cert failures, last attempt " .. math.floor(elapsed) .. "s ago, want to wait " .. backoff .. "s before trying again")
          return false
        end

Conceptually, the simplest mechanism to me would be one that (a) caches the results of `allow_domain`, (b) equally takes into consideration cert failures, and (c) uses permanent storage so that the tracking automatically shared across multiple servers. (My patch only does (b) and gives the implementer the tools to implement (a).)

Reply:
Another approach that I didn't attempt, but which would make sense to me also: track each certificate request (success and failure), have one tunable to set the max number of certificate requests per three hours (defaulting to LE's `300`), and have algorithms to adaptively throttle both renewals, and also (failing) ad-hoc certificate requests as needed. With additional tunable as needed, and using permanent storage (i.e. aggregated across multiple servers).